lib/Semantic_check.ml

```diff
@@ -1,2163 +0,0 @@
-  | Identity -> Identity
-  | Lower e ->
-      let ue = semantic_check_expression e in
-      let loc = (snd (typed_expression_unroll ue)).expr_typed_meta_loc in
-      let _ =
-        if not (check_of_int_or_real_type ue) then
-          semantic_error ~loc
-            ( "Lower bound should be of int or real type. Instead found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue)
-            ^ "." )
-      in
-      Lower ue
-  | Upper e ->
-      let ue = semantic_check_expression e in
-      let loc = (snd (typed_expression_unroll ue)).expr_typed_meta_loc in
-      let _ =
-        if not (check_of_int_or_real_type ue) then
-          semantic_error ~loc
-            ( "Upper bound should be of int or real type. Instead found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue)
-            ^ "." )
-      in
-      Upper ue
-  | LowerUpper (e1, e2) ->
-      let ue1 = semantic_check_expression e1 in
-      let ue2 = semantic_check_expression e2 in
-      let loc1 = (snd (typed_expression_unroll ue1)).expr_typed_meta_loc in
-      let loc2 = (snd (typed_expression_unroll ue2)).expr_typed_meta_loc in
-      let _ =
-        if not (check_of_int_or_real_type ue1) then
-          semantic_error ~loc:loc1
-            ( "Lower and upper bound should be of int or real type. Instead \
-               found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue1)
-            ^ "." )
-      in
-      let _ =
-        if not (check_of_int_or_real_type ue2) then
-          semantic_error ~loc:loc2
-            ( "Lower and upper bound should be of int or real type. Instead \
-               found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue2)
-            ^ "." )
-      in
-      LowerUpper (ue1, ue2)
-  | LocationScale (e1, e2) ->
-      let ue1 = semantic_check_expression e1 in
-      let ue2 = semantic_check_expression e2 in
-      let loc1 = (snd (typed_expression_unroll ue1)).expr_typed_meta_loc in
-      let loc2 = (snd (typed_expression_unroll ue2)).expr_typed_meta_loc in
-      let _ =
-        if not (check_of_int_or_real_type ue1) then
-          semantic_error ~loc:loc1
-            ( "Location and scale should be of int or real type. Instead \
-               found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue1)
-            ^ "." )
-      in
-      let _ =
-        if not (check_of_int_or_real_type ue2) then
-          semantic_error ~loc:loc2
-            ( "Location and scale should be of int or real type. Instead \
-               found type "
-            ^ pretty_print_unsizedtype (type_of_typed_expr ue2)
-            ^ "." )
-      in
-      LocationScale (ue1, ue2)
-  | Ordered -> Ordered
-  | PositiveOrdered -> PositiveOrdered
-  | Simplex -> Simplex
-  | UnitVector -> UnitVector
-  | CholeskyCorr -> CholeskyCorr
-  | CholeskyCov -> CholeskyCov
-  | Correlation -> Correlation
-  | Covariance -> Covariance
```

Comment:
Isn't this all summarized as `x -> x`?

Reply:
Yes! Done.

Reply:
Ah. Actually, we can't, as we're going from an untyped to a typed AST (which happen to share many labels). The type checker can't figure it out if you write x -> x.

Reply:
These transformations or constraints or whatever we call them are not shared? Could we share them?

Reply:
They are shared, but for some reason the type checker complains if you write x -> x.